diff --git a/glamor/Makefile.am b/glamor/Makefile.am
index aaf0aab..a7354c9 100644
--- a/glamor/Makefile.am
+++ b/glamor/Makefile.am
@@ -1,8 +1,13 @@
-noinst_LTLIBRARIES = libglamor.la libglamor_egl_stubs.la
+noinst_LTLIBRARIES = libglamor.la libglamor_egl.la libglamor_egl_stubs.la
 
 libglamor_la_LIBADD = $(GLAMOR_LIBS)
 
-AM_CFLAGS = $(CWARNFLAGS) $(DIX_CFLAGS) $(GLAMOR_CFLAGS)
+AM_CFLAGS = \
+       -I$(top_srcdir)/dri3 \
+       $(CWARNFLAGS) \
+       $(DIX_CFLAGS) \
+       $(GLAMOR_CFLAGS) \
+       $(GBM_CFLAGS) $(DRM_CFLAGS) $(EPOXY_CFLAGS)
 
 libglamor_la_SOURCES = \
 	glamor.c \
@@ -54,6 +59,15 @@ libglamor_la_SOURCES += \
 	glamor_xv.c
 endif
 
+libglamor_egl_la_SOURCES = \
+       glamor_egl.c
+
+libglamor_egl_la_LIBADD = \
+        $(GLAMOR_EGL_LIBS) \
+        $(GBM_LIBS) \
+        $(DRM_LIBS) \
+        $(EPOXY_LIBS)
+
 libglamor_egl_stubs_la_SOURCES = \
 	glamor_egl_stubs.c \
 	glamor_egl_ext.h \
diff --git a/glamor/glamor_egl.c b/glamor/glamor_egl.c
index 525b1f4..8dea271 100644
--- a/glamor/glamor_egl.c
+++ b/glamor/glamor_egl.c
@@ -29,16 +29,13 @@
 
 #include "dix-config.h"
 
-#define GLAMOR_FOR_XORG
 #include <unistd.h>
 #include <fcntl.h>
 #include <sys/ioctl.h>
 #include <errno.h>
-#include <xf86.h>
-#include <xf86Priv.h>
-#include <xf86drm.h>
 #define EGL_DISPLAY_NO_X_MESA
 
+#include <xf86drm.h>
 #include <gbm.h>
 #include <drm_fourcc.h>
 
@@ -47,6 +44,18 @@
 #include "glamor.h"
 #include "glamor_priv.h"
 #include "dri3.h"
+#ifdef GLAMOR_FOR_XORG
+#include <xf86.h>
+#include <xf86Priv.h>
+#endif
+
+static Bool
+glamor_egl_create_textured_pixmap(PixmapPtr pixmap, int handle, int stride);
+
+_X_EXPORT Bool
+glamor_egl_create_textured_pixmap_from_gbm_bo(PixmapPtr pixmap,
+                                              struct gbm_bo *bo,
+                                              Bool used_modifiers);
 
 struct glamor_egl_screen_private {
     EGLDisplay display;
@@ -61,9 +70,12 @@ struct glamor_egl_screen_private {
 
     CloseScreenProcPtr saved_close_screen;
     DestroyPixmapProcPtr saved_destroy_pixmap;
+#ifdef GLAMOR_FOR_XORG
     xf86FreeScreenProc *saved_free_screen;
+#endif
 };
 
+#ifdef GLAMOR_FOR_XORG
 int xf86GlamorEGLPrivateIndex = -1;
 
 
@@ -73,6 +85,16 @@ glamor_egl_get_screen_private(ScrnInfoPtr scrn)
     return (struct glamor_egl_screen_private *)
         scrn->privates[xf86GlamorEGLPrivateIndex].ptr;
 }
+#else
+static DevPrivateKeyRec glamor_egl_screen_private_key;
+
+static struct glamor_egl_screen_private *
+glamor_egl_get_screen_private(ScreenPtr screen)
+{
+    return dixLookupPrivate(&screen->devPrivates,
+                            &glamor_egl_screen_private_key);
+}
+#endif
 
 static void
 glamor_egl_make_current(struct glamor_context *glamor_ctx)
@@ -101,15 +123,15 @@ glamor_get_flink_name(int fd, int handle, int *name)
     flink.handle = handle;
     if (ioctl(fd, DRM_IOCTL_GEM_FLINK, &flink) < 0) {
 
-	/*
-	 * Assume non-GEM kernels have names identical to the handle
-	 */
-	if (errno == ENODEV) {
-	    *name = handle;
-	    return TRUE;
-	} else {
-	    return FALSE;
-	}
+        /*
+         * Assume non-GEM kernels have names identical to the handle
+         */
+        if (errno == ENODEV) {
+            *name = handle;
+            return TRUE;
+        } else {
+            return FALSE;
+        }
     }
     *name = flink.name;
     return TRUE;
@@ -138,22 +160,33 @@ glamor_create_texture_from_image(ScreenPtr screen,
 struct gbm_device *
 glamor_egl_get_gbm_device(ScreenPtr screen)
 {
+#ifdef GLAMOR_FOR_XORG
     struct glamor_egl_screen_private *glamor_egl =
         glamor_egl_get_screen_private(xf86ScreenToScrn(screen));
+#else
+    struct glamor_egl_screen_private *glamor_egl =
+        glamor_egl_get_screen_private(screen);
+#endif
     return glamor_egl->gbm;
 }
 
 Bool
 glamor_egl_create_textured_screen(ScreenPtr screen, int handle, int stride)
 {
+#ifdef GLAMOR_FOR_XORG
     ScrnInfoPtr scrn = xf86ScreenToScrn(screen);
+#endif
     PixmapPtr screen_pixmap;
 
     screen_pixmap = screen->GetScreenPixmap(screen);
 
     if (!glamor_egl_create_textured_pixmap(screen_pixmap, handle, stride)) {
+#ifdef GLAMOR_FOR_XORG
         xf86DrvMsg(scrn->scrnIndex, X_ERROR,
                    "Failed to create textured screen.");
+#else
+        LogMessage(X_ERROR, "Failed to create textured screen.");
+#endif
         return FALSE;
     }
     return TRUE;
@@ -170,22 +203,30 @@ glamor_egl_set_pixmap_image(PixmapPtr pixmap, EGLImageKHR image,
     old = pixmap_priv->image;
     if (old) {
         ScreenPtr                               screen = pixmap->drawable.pScreen;
-        ScrnInfoPtr                             scrn = xf86ScreenToScrn(screen);
-        struct glamor_egl_screen_private        *glamor_egl = glamor_egl_get_screen_private(scrn);
-
+#ifdef GLAMOR_FOR_XORG
+        struct glamor_egl_screen_private        *glamor_egl =
+            glamor_egl_get_screen_private(xf86ScreenToScrn(screen));
+#else
+        struct glamor_egl_screen_private        *glamor_egl =
+            glamor_egl_get_screen_private(screen);
+#endif
         eglDestroyImageKHR(glamor_egl->display, old);
     }
     pixmap_priv->image = image;
     pixmap_priv->used_modifiers = used_modifiers;
 }
 
-Bool
+static Bool
 glamor_egl_create_textured_pixmap(PixmapPtr pixmap, int handle, int stride)
 {
     ScreenPtr screen = pixmap->drawable.pScreen;
-    ScrnInfoPtr scrn = xf86ScreenToScrn(screen);
+#ifdef GLAMOR_FOR_XORG
     struct glamor_egl_screen_private *glamor_egl =
-        glamor_egl_get_screen_private(scrn);
+        glamor_egl_get_screen_private(xf86ScreenToScrn(screen));
+#else
+    struct glamor_egl_screen_private *glamor_egl =
+        glamor_egl_get_screen_private(screen);
+#endif
     int ret, fd;
 
     /* GBM doesn't have an import path from handles, so we make a
@@ -193,19 +234,29 @@ glamor_egl_create_textured_pixmap(PixmapPtr pixmap, int handle, int stride)
      */
     ret = drmPrimeHandleToFD(glamor_egl->fd, handle, O_CLOEXEC, &fd);
     if (ret) {
+#ifdef GLAMOR_FOR_XORG
+        ScrnInfoPtr scrn = xf86ScreenToScrn(screen);
         xf86DrvMsg(scrn->scrnIndex, X_ERROR,
                    "Failed to make prime FD for handle: %d\n", errno);
+#else
+        LogMessage(X_ERROR, "Failed to make prime FD for handle: %d\n", errno);
+#endif
         return FALSE;
     }
 
     if (!glamor_back_pixmap_from_fd(pixmap, fd,
                                     pixmap->drawable.width,
                                     pixmap->drawable.height,
-                                    stride,
-                                    pixmap->drawable.depth,
+                                    stride, pixmap->drawable.depth,
                                     pixmap->drawable.bitsPerPixel)) {
+#ifdef GLAMOR_FOR_XORG
+        ScrnInfoPtr scrn = xf86ScreenToScrn(screen);
         xf86DrvMsg(scrn->scrnIndex, X_ERROR,
                    "Failed to make import prime FD as pixmap: %d\n", errno);
+#else
+        LogMessage(X_ERROR,
+                   "Failed to make import prime FD as pixmap: %d\n", errno);
+#endif
         close(fd);
         return FALSE;
     }
@@ -214,21 +265,24 @@ glamor_egl_create_textured_pixmap(PixmapPtr pixmap, int handle, int stride)
     return TRUE;
 }
 
-Bool
+_X_EXPORT Bool
 glamor_egl_create_textured_pixmap_from_gbm_bo(PixmapPtr pixmap,
                                               struct gbm_bo *bo,
                                               Bool used_modifiers)
 {
     ScreenPtr screen = pixmap->drawable.pScreen;
-    ScrnInfoPtr scrn = xf86ScreenToScrn(screen);
     struct glamor_screen_private *glamor_priv =
         glamor_get_screen_private(screen);
     struct glamor_egl_screen_private *glamor_egl;
     EGLImageKHR image;
     GLuint texture;
     Bool ret = FALSE;
-
-    glamor_egl = glamor_egl_get_screen_private(scrn);
+#ifdef GLAMOR_FOR_XORG
+    glamor_egl =
+        glamor_egl_get_screen_private(xf86ScreenToScrn(screen));
+#else
+    glamor_egl = glamor_egl_get_screen_private(screen);
+#endif
 
     glamor_make_current(glamor_priv);
 
@@ -263,9 +317,13 @@ static Bool
 glamor_make_pixmap_exportable(PixmapPtr pixmap, Bool modifiers_ok)
 {
     ScreenPtr screen = pixmap->drawable.pScreen;
-    ScrnInfoPtr scrn = xf86ScreenToScrn(screen);
+#ifdef GLAMOR_FOR_XORG
     struct glamor_egl_screen_private *glamor_egl =
-        glamor_egl_get_screen_private(scrn);
+        glamor_egl_get_screen_private(xf86ScreenToScrn(screen));
+#else
+    struct glamor_egl_screen_private *glamor_egl =
+        glamor_egl_get_screen_private(screen);
+#endif
     struct glamor_pixmap_private *pixmap_priv =
         glamor_get_pixmap_private(pixmap);
     unsigned width = pixmap->drawable.width;
@@ -298,9 +356,15 @@ glamor_make_pixmap_exportable(PixmapPtr pixmap, Bool modifiers_ok)
         format = GBM_FORMAT_R8;
         break;
     default:
+#ifdef GLAMOR_FOR_XORG
+        ScrnInfoPtr scrn = xf86ScreenToScrn(screen);
         xf86DrvMsg(scrn->scrnIndex, X_ERROR,
                    "Failed to make %d depth, %dbpp pixmap exportable\n",
                    pixmap->drawable.depth, pixmap->drawable.bitsPerPixel);
+#else
+        LogMessage(X_ERROR, "Failed to make %d depth, %dbpp pixmap exportable\n",
+                   pixmap->drawable.depth, pixmap->drawable.bitsPerPixel);
+#endif
         return FALSE;
     }
 
@@ -330,20 +394,32 @@ glamor_make_pixmap_exportable(PixmapPtr pixmap, Bool modifiers_ok)
     }
 
     if (!bo) {
+#ifdef GLAMOR_FOR_XORG
+        ScrnInfoPtr scrn = xf86ScreenToScrn(screen);
         xf86DrvMsg(scrn->scrnIndex, X_ERROR,
                    "Failed to make %dx%dx%dbpp GBM bo\n",
                    width, height, pixmap->drawable.bitsPerPixel);
+#else
+        LogMessage(X_ERROR, "Failed to make %dx%dx%dbpp GBM bo\n",
+                   width, height, pixmap->drawable.bitsPerPixel);
+#endif
         return FALSE;
     }
 
     exported = screen->CreatePixmap(screen, 0, 0, pixmap->drawable.depth, 0);
     screen->ModifyPixmapHeader(exported, width, height, 0, 0,
                                gbm_bo_get_stride(bo), NULL);
-    if (!glamor_egl_create_textured_pixmap_from_gbm_bo(exported, bo,
-                                                       used_modifiers)) {
+    if (!glamor_egl_create_textured_pixmap_from_gbm_bo
+        (exported, bo, used_modifiers)) {
+#ifdef GLAMOR_FOR_XORG
+        ScrnInfoPtr scrn = xf86ScreenToScrn(screen);
         xf86DrvMsg(scrn->scrnIndex, X_ERROR,
                    "Failed to make %dx%dx%dbpp pixmap from GBM bo\n",
                    width, height, pixmap->drawable.bitsPerPixel);
+#else
+        LogMessage(X_ERROR, "Failed to make %dx%dx%dbpp pixmap from GBM bo\n",
+                   width, height, pixmap->drawable.bitsPerPixel);
+#endif
         screen->DestroyPixmap(exported);
         gbm_bo_destroy(bo);
         return FALSE;
@@ -373,8 +449,13 @@ glamor_make_pixmap_exportable(PixmapPtr pixmap, Bool modifiers_ok)
 static struct gbm_bo *
 glamor_gbm_bo_from_pixmap_internal(ScreenPtr screen, PixmapPtr pixmap)
 {
+#ifdef GLAMOR_FOR_XORG
     struct glamor_egl_screen_private *glamor_egl =
         glamor_egl_get_screen_private(xf86ScreenToScrn(screen));
+#else
+    struct glamor_egl_screen_private *glamor_egl =
+        glamor_egl_get_screen_private(screen);
+#endif
     struct glamor_pixmap_private *pixmap_priv =
         glamor_get_pixmap_private(pixmap);
 
@@ -432,7 +513,7 @@ glamor_egl_fds_from_pixmap(ScreenPtr screen, PixmapPtr pixmap, int *fds,
          * gbm_bo_get_fd without losing information. If they point to different
          * objects we are out of luck and need to give up.
          */
-	if (first_handle == plane_handle.s32)
+        if (first_handle == plane_handle.s32)
             fds[i] = gbm_bo_get_fd(bo);
         else
             fds[i] = -1;
@@ -498,7 +579,11 @@ glamor_egl_fd_name_from_pixmap(ScreenPtr screen,
     struct gbm_bo *bo;
     int fd = -1;
 
+#ifdef GLAMOR_FOR_XORG
     glamor_egl = glamor_egl_get_screen_private(xf86ScreenToScrn(screen));
+#else
+    glamor_egl = glamor_egl_get_screen_private(screen);
+#endif
 
     if (!glamor_make_pixmap_exportable(pixmap, FALSE))
         goto failure;
@@ -551,14 +636,15 @@ glamor_back_pixmap_from_fd(PixmapPtr pixmap,
                            CARD16 stride, CARD8 depth, CARD8 bpp)
 {
     ScreenPtr screen = pixmap->drawable.pScreen;
-    ScrnInfoPtr scrn = xf86ScreenToScrn(screen);
-    struct glamor_egl_screen_private *glamor_egl;
+#ifdef GLAMOR_FOR_XORG
+    struct glamor_egl_screen_private *glamor_egl = glamor_egl_get_screen_private(xf86ScreenToScrn(screen));
+#else
+    struct glamor_egl_screen_private *glamor_egl = glamor_egl_get_screen_private(screen);
+#endif
     struct gbm_bo *bo;
     struct gbm_import_fd_data import_data = { 0 };
     Bool ret;
 
-    glamor_egl = glamor_egl_get_screen_private(scrn);
-
     if (!gbm_format_for_depth(depth, &import_data.format) ||
         width == 0 || height == 0)
         return FALSE;
@@ -591,7 +677,11 @@ glamor_pixmap_from_fds(ScreenPtr screen,
     Bool ret = FALSE;
     int i;
 
+#ifdef GLAMOR_FOR_XORG
     glamor_egl = glamor_egl_get_screen_private(xf86ScreenToScrn(screen));
+#else
+    glamor_egl = glamor_egl_get_screen_private(screen);
+#endif
 
     pixmap = screen->CreatePixmap(screen, 0, 0, depth, 0);
 
@@ -669,7 +759,11 @@ glamor_get_formats(ScreenPtr screen,
     /* Explicitly zero the count as the caller may ignore the return value */
     *num_formats = 0;
 
+#ifdef GLAMOR_FOR_XORG
     glamor_egl = glamor_egl_get_screen_private(xf86ScreenToScrn(screen));
+#else
+    glamor_egl = glamor_egl_get_screen_private(screen);
+#endif
 
     if (!glamor_egl->dmabuf_capable)
         return TRUE;
@@ -709,7 +803,11 @@ glamor_get_modifiers(ScreenPtr screen, uint32_t format,
     /* Explicitly zero the count as the caller may ignore the return value */
     *num_modifiers = 0;
 
+#ifdef GLAMOR_FOR_XORG
     glamor_egl = glamor_egl_get_screen_private(xf86ScreenToScrn(screen));
+#else
+    glamor_egl = glamor_egl_get_screen_private(screen);
+#endif
 
     if (!glamor_egl->dmabuf_capable)
         return FALSE;
@@ -740,6 +838,7 @@ glamor_get_modifiers(ScreenPtr screen, uint32_t format,
 #endif
 }
 
+#ifdef GLAMOR_FOR_XORG
 _X_EXPORT const char *
 glamor_egl_get_driver_name(ScreenPtr screen)
 {
@@ -754,15 +853,19 @@ glamor_egl_get_driver_name(ScreenPtr screen)
 
     return NULL;
 }
-
+#endif
 
 static Bool
 glamor_egl_destroy_pixmap(PixmapPtr pixmap)
 {
     ScreenPtr screen = pixmap->drawable.pScreen;
-    ScrnInfoPtr scrn = xf86ScreenToScrn(screen);
+#ifdef GLAMOR_FOR_XORG
     struct glamor_egl_screen_private *glamor_egl =
-        glamor_egl_get_screen_private(scrn);
+        glamor_egl_get_screen_private(xf86ScreenToScrn(screen));
+#else
+    struct glamor_egl_screen_private *glamor_egl =
+        glamor_egl_get_screen_private(screen);
+#endif
     Bool ret;
 
     if (pixmap->refcnt == 1) {
@@ -807,21 +910,33 @@ glamor_egl_exchange_buffers(PixmapPtr front, PixmapPtr back)
 static Bool
 glamor_egl_close_screen(ScreenPtr screen)
 {
-    ScrnInfoPtr scrn;
     struct glamor_egl_screen_private *glamor_egl;
     struct glamor_pixmap_private *pixmap_priv;
     PixmapPtr screen_pixmap;
 
-    scrn = xf86ScreenToScrn(screen);
-    glamor_egl = glamor_egl_get_screen_private(scrn);
+#ifdef GLAMOR_FOR_XORG
+    glamor_egl = glamor_egl_get_screen_private(xf86ScreenToScrn(screen));
+#else
+    glamor_egl = glamor_egl_get_screen_private(screen);
+#endif
     screen_pixmap = screen->GetScreenPixmap(screen);
     pixmap_priv = glamor_get_pixmap_private(screen_pixmap);
 
-    eglDestroyImageKHR(glamor_egl->display, pixmap_priv->image);
-    pixmap_priv->image = NULL;
-
+    if (pixmap_priv && pixmap_priv->image) {
+        eglDestroyImageKHR(glamor_egl->display, pixmap_priv->image);
+        pixmap_priv->image = NULL;
+    }
     screen->CloseScreen = glamor_egl->saved_close_screen;
 
+#ifndef GLAMOR_FOR_XORG
+    if (glamor_egl->display != EGL_NO_DISPLAY) {
+        eglMakeCurrent(glamor_egl->display,
+                       EGL_NO_SURFACE, EGL_NO_SURFACE, EGL_NO_CONTEXT);
+        lastGLContext = NULL;
+        eglTerminate(glamor_egl->display);
+    }
+#endif
+
     return screen->CloseScreen(screen);
 }
 
@@ -832,9 +949,13 @@ glamor_dri3_open_client(ClientPtr client,
                         RRProviderPtr provider,
                         int *fdp)
 {
-    ScrnInfoPtr scrn = xf86ScreenToScrn(screen);
+#ifdef GLAMOR_FOR_XORG
     struct glamor_egl_screen_private *glamor_egl =
-        glamor_egl_get_screen_private(scrn);
+        glamor_egl_get_screen_private(xf86ScreenToScrn(screen));
+#else
+    struct glamor_egl_screen_private *glamor_egl =
+        glamor_egl_get_screen_private(screen);
+#endif
     int fd;
     drm_magic_t magic;
 
@@ -891,9 +1012,13 @@ static const dri3_screen_info_rec glamor_dri3_info = {
 void
 glamor_egl_screen_init(ScreenPtr screen, struct glamor_context *glamor_ctx)
 {
-    ScrnInfoPtr scrn = xf86ScreenToScrn(screen);
+#ifdef GLAMOR_FOR_XORG
     struct glamor_egl_screen_private *glamor_egl =
-        glamor_egl_get_screen_private(scrn);
+        glamor_egl_get_screen_private(xf86ScreenToScrn(screen));
+#else
+    struct glamor_egl_screen_private *glamor_egl =
+        glamor_egl_get_screen_private(screen);
+#endif
 #ifdef DRI3
     glamor_screen_private *glamor_priv = glamor_get_screen_private(screen);
 #endif
@@ -927,8 +1052,13 @@ glamor_egl_screen_init(ScreenPtr screen, struct glamor_context *glamor_ctx)
         glamor_egl->device_path = drmGetDeviceNameFromFd2(glamor_egl->fd);
 
         if (!dri3_screen_init(screen, &glamor_dri3_info)) {
+#ifdef GLAMOR_FOR_XORG
+            ScrnInfoPtr scrn = xf86ScreenToScrn(screen);
             xf86DrvMsg(scrn->scrnIndex, X_ERROR,
                        "Failed to initialize DRI3.\n");
+#else
+            LogMessage(X_ERROR, "Failed to initialize DRI3.\n");
+#endif
         }
     }
 #endif
@@ -952,6 +1082,7 @@ static void glamor_egl_cleanup(struct glamor_egl_screen_private *glamor_egl)
     free(glamor_egl);
 }
 
+#ifdef GLAMOR_FOR_XORG
 static void
 glamor_egl_free_screen(ScrnInfoPtr scrn)
 {
@@ -1164,29 +1295,41 @@ glamor_egl_init(ScrnInfoPtr scrn, int fd)
     lastGLContext = NULL;
 
     if (!epoxy_has_gl_extension("GL_OES_EGL_image")) {
+#ifdef GLAMOR_FOR_XORG
         xf86DrvMsg(scrn->scrnIndex, X_ERROR,
                    "glamor acceleration requires GL_OES_EGL_image\n");
+#else
+        LogMessage(X_ERROR, "glamor acceleration requires GL_OES_EGL_image\n");
+#endif
         goto error;
     }
 
+#ifdef GLAMOR_FOR_XORG
     xf86DrvMsg(scrn->scrnIndex, X_INFO, "glamor X acceleration enabled on %s\n",
                renderer);
+#endif
 
 #ifdef GBM_BO_WITH_MODIFIERS
     if (epoxy_has_egl_extension(glamor_egl->display,
                                 "EGL_EXT_image_dma_buf_import") &&
         epoxy_has_egl_extension(glamor_egl->display,
                                 "EGL_EXT_image_dma_buf_import_modifiers")) {
+#ifdef GLAMOR_FOR_XORG
         if (xf86Info.debug != NULL)
-            glamor_egl->dmabuf_capable = !!strstr(xf86Info.debug,
-                                                  "dmabuf_capable");
+            glamor_egl->dmabuf_capable =
+                !!strstr(xf86Info.debug, "dmabuf_capable");
         else
             glamor_egl->dmabuf_capable = FALSE;
+#else
+        glamor_egl->dmabuf_capable = TRUE;
+#endif
     }
 #endif
 
+#ifdef GLAMOR_FOR_XORG
     glamor_egl->saved_free_screen = scrn->FreeScreen;
     scrn->FreeScreen = glamor_egl_free_screen;
+#endif
     return TRUE;
 
 error:
@@ -1200,3 +1343,143 @@ glamor_egl_init_textured_pixmap(ScreenPtr screen)
 {
     return TRUE;
 }
+#else
+_X_EXPORT Bool
+glamor_egl_init(ScreenPtr screen, int fd)
+{
+    struct glamor_egl_screen_private *glamor_egl;
+    const GLubyte *renderer;
+    EGLConfig egl_config;
+    int n;
+
+    glamor_egl = calloc(1, sizeof(*glamor_egl));
+    if (glamor_egl == NULL)
+        return FALSE;
+
+    if (!dixRegisterPrivateKey(&glamor_egl_screen_private_key,
+                               PRIVATE_SCREEN, 0)) {
+        free(glamor_egl);
+        return FALSE;
+    }
+    dixSetPrivate(&screen->devPrivates, &glamor_egl_screen_private_key,
+                  glamor_egl);
+
+    glamor_egl->fd = fd;
+    glamor_egl->context = EGL_NO_CONTEXT;
+    glamor_egl->display = EGL_NO_DISPLAY;
+    glamor_egl->gbm = gbm_create_device(fd);
+    if (glamor_egl->gbm == NULL) {
+        LogMessage(X_ERROR, "couldn't create gbm device\n");
+        goto error;
+    }
+
+    glamor_egl->display = eglGetPlatformDisplayEXT(EGL_PLATFORM_GBM_KHR,
+                                                    glamor_egl->gbm, NULL);
+    if (glamor_egl->display == EGL_NO_DISPLAY) {
+        LogMessage(X_ERROR,
+                   "eglGetPlatformDisplayEXT() failed with error %d\n",
+                   eglGetError());
+        goto error;
+    }
+
+    if (!eglInitialize(glamor_egl->display, NULL, NULL)) {
+        LogMessage(X_ERROR, "eglInitialize() failed\n");
+        glamor_egl->display = EGL_NO_DISPLAY;
+        goto error;
+    }
+
+    if (!epoxy_has_egl_extension(glamor_egl->display,
+                                 "EGL_KHR_surfaceless_context")) {
+        LogMessage(X_ERROR, "EGL_KHR_surfaceless_context required.\n");
+        goto error;
+    }
+
+    if (eglBindAPI(EGL_OPENGL_API)) {
+        static const EGLint context_attribs_core[] = {
+            EGL_CONTEXT_OPENGL_PROFILE_MASK_KHR,
+            EGL_CONTEXT_OPENGL_CORE_PROFILE_BIT_KHR,
+            EGL_CONTEXT_MAJOR_VERSION_KHR,
+            GLAMOR_GL_CORE_VER_MAJOR,
+            EGL_CONTEXT_MINOR_VERSION_KHR,
+            GLAMOR_GL_CORE_VER_MINOR,
+            EGL_NONE
+        };
+        static const EGLint context_attribs[] = { EGL_NONE };
+
+        glamor_egl->context = eglCreateContext(glamor_egl->display,
+                                               EGL_NO_CONFIG_KHR,
+                                               EGL_NO_CONTEXT,
+                                               context_attribs_core);
+        if (glamor_egl->context == EGL_NO_CONTEXT)
+            glamor_egl->context = eglCreateContext(glamor_egl->display,
+                                                   EGL_NO_CONFIG_KHR,
+                                                   EGL_NO_CONTEXT,
+                                                   context_attribs);
+    }
+
+    if (glamor_egl->context == EGL_NO_CONTEXT) {
+        static const EGLint context_attribs[] = {
+            EGL_CONTEXT_CLIENT_VERSION, 2,
+            EGL_NONE
+        };
+
+        if (!eglBindAPI(EGL_OPENGL_ES_API)) {
+            LogMessage(X_ERROR,
+                       "glamor: Failed to bind either GL or GLES APIs.\n");
+            goto error;
+        }
+        if (!eglChooseConfig(glamor_egl->display, NULL, &egl_config, 1, &n) ||
+            n < 1) {
+            LogMessage(X_ERROR, "glamor: No acceptable EGL configs found\n");
+            goto error;
+        }
+        glamor_egl->context = eglCreateContext(glamor_egl->display,
+                                               egl_config, EGL_NO_CONTEXT,
+                                               context_attribs);
+    }
+
+    if (glamor_egl->context == EGL_NO_CONTEXT) {
+        LogMessage(X_ERROR,
+                   "glamor: Failed to create GL or GLES2 contexts\n");
+        goto error;
+    }
+    if (!eglMakeCurrent(glamor_egl->display, EGL_NO_SURFACE,
+                        EGL_NO_SURFACE, glamor_egl->context)) {
+        LogMessage(X_ERROR, "Failed to make EGL context current\n");
+        goto error;
+    }
+
+    renderer = glGetString(GL_RENDERER);
+    if (!renderer) {
+        LogMessage(X_ERROR,
+                   "glGetString() returned NULL, your GL is broken\n");
+        goto error;
+    }
+    if (strstr((const char *) renderer, "llvmpipe")) {
+        LogMessage(X_INFO, "Refusing to try glamor on llvmpipe\n");
+        goto error;
+    }
+
+    lastGLContext = NULL;
+    if (!epoxy_has_gl_extension("GL_OES_EGL_image")) {
+        LogMessage(X_ERROR,
+                   "glamor acceleration requires GL_OES_EGL_image\n");
+        goto error;
+    }
+
+#ifdef GBM_BO_WITH_MODIFIERS
+    if (epoxy_has_egl_extension(glamor_egl->display,
+                                "EGL_EXT_image_dma_buf_import") &&
+        epoxy_has_egl_extension(glamor_egl->display,
+                                "EGL_EXT_image_dma_buf_import_modifiers"))
+        glamor_egl->dmabuf_capable = TRUE;
+#endif
+
+    return TRUE;
+
+error:
+    dixSetPrivate(&screen->devPrivates, &glamor_egl_screen_private_key, NULL);
+    glamor_egl_cleanup(glamor_egl);
+    return FALSE;
+}
+#endif
diff --git a/glamor/glamor_egl.h b/glamor/glamor_egl.h
index 8f6ed78..5edc016 100644
--- a/glamor/glamor_egl.h
+++ b/glamor/glamor_egl.h
@@ -27,6 +27,12 @@
 #ifndef GLAMOR_EGL_H
 #define GLAMOR_EGL_H
 
+#include <scrnintstr.h>
+#include <pixmapstr.h>
+
+struct gbm_bo;
+struct glamor_context;
+
 #define MESA_EGL_NO_X11_HEADERS
 #define EGL_NO_X11
 #include <epoxy/gl.h>
@@ -74,4 +80,37 @@ glamor_egl_get_display(EGLint type, void *native)
     return eglGetDisplay(native);
 }
 
+#ifdef GLAMOR_FOR_XORG
+
+#define GLAMOR_EGL_MODULE_NAME  "glamoregl"
+
+extern _X_EXPORT Bool glamor_egl_init(ScrnInfoPtr scrn, int fd);
+
+extern _X_EXPORT Bool glamor_egl_init_textured_pixmap(ScreenPtr screen);
+
+extern _X_EXPORT Bool glamor_egl_create_textured_screen(ScreenPtr screen,
+                                                        int handle, int stride);
+
+extern _X_EXPORT Bool glamor_egl_create_textured_pixmap(PixmapPtr pixmap,
+                                                        int handle, int stride);
+
+extern _X_EXPORT const char *glamor_egl_get_driver_name(ScreenPtr screen);
+
+#else
+/*
+ * These are the glamor EGL interfaces required for DIX-only servers
+ * like Xvfb that do not use the full Xorg DDX framework.
+ */
+extern _X_EXPORT Bool glamor_egl_init(ScreenPtr screen, int fd);
 #endif
+
+/* This function is used by both Xorg DDX drivers and Xvfb. */
+extern _X_EXPORT Bool
+glamor_egl_create_textured_pixmap_from_gbm_bo(PixmapPtr pixmap,
+                                              struct gbm_bo *bo,
+                                              Bool used_modifiers);
+
+extern _X_EXPORT void glamor_egl_screen_init(ScreenPtr screen,
+                                             struct glamor_context *glamor_ctx);
+
+#endif /* GLAMOR_EGL_H */
diff --git a/glamor/meson.build b/glamor/meson.build
index 268af59..ffbd302 100644
--- a/glamor/meson.build
+++ b/glamor/meson.build
@@ -40,12 +40,18 @@ endif
 
 epoxy_dep = dependency('epoxy')
 
+if get_option('glamor') == 'true'
+    gbm_dep = dependency('gbm', version: '>= 17.1.0')
+    drm_dep = dependency('libdrm')
+endif
+
 glamor = static_library('glamor',
     srcs_glamor,
     include_directories: inc,
     dependencies: [
         common_dep,
         epoxy_dep,
+        gbm_dep,
     ],
 )
 
@@ -55,6 +61,18 @@ glamor_egl_stubs = static_library('glamor_egl_stubs',
     dependencies: common_dep,
 )
 
+glamor_egl = static_library('glamor_egl',
+    'glamor_egl.c',
+    include_directories: inc,
+    dependencies: [
+        common_dep,
+        epoxy_dep,
+        gbm_dep,
+        drm_dep,
+    ],
+    install: false,
+)
+
 if build_xorg
     install_data('glamor.h', install_dir: xorgsdkdir)
 endif
diff --git a/hw/vfb/InitOutput.c b/hw/vfb/InitOutput.c
index 48efb61..5475c28 100644
--- a/hw/vfb/InitOutput.c
+++ b/hw/vfb/InitOutput.c
@@ -55,6 +55,7 @@ from The Open Group.
 #endif                          /* HAVE_MMAP */
 #include <sys/stat.h>
 #include <errno.h>
+static char *vfbDeviceNode = NULL;
 #ifndef WIN32
 #include <sys/param.h>
 #endif
@@ -68,6 +69,15 @@ from The Open Group.
 #include "glx_extinit.h"
 #include "randrstr.h"
 
+#ifdef GLAMOR_HAS_GBM
+#include <glamor.h>
+#include <glamor_egl.h>
+#include <gbm.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <errno.h>
+#endif
+
 #define VFB_DEFAULT_WIDTH      1280
 #define VFB_DEFAULT_HEIGHT     1024
 #define VFB_DEFAULT_DEPTH        24
@@ -101,6 +111,12 @@ typedef struct {
 #ifdef HAS_SHM
     int shmid;
 #endif
+#ifdef GLAMOR_HAS_GBM
+    int fd;
+    CreateScreenResourcesProcPtr createScreenResources;
+    struct gbm_device *gbm;
+    struct gbm_bo *front_bo;
+#endif
 } vfbScreenInfo, *vfbScreenInfoPtr;
 
 static int vfbNumScreens;
@@ -255,6 +271,9 @@ ddxUseMsg(void)
 #ifdef HAS_SHM
     ErrorF("-shmem                 put framebuffers in shared memory\n");
 #endif
+#ifdef GLAMOR_HAS_GBM
+    ErrorF("-vfbdevice device-path use specified DRI render node for glamor/DRI3\n");
+#endif
 }
 
 int
@@ -375,6 +394,14 @@ ddxProcessArgument(int argc, char *argv[], int i)
     }
 #endif
 
+#ifdef GLAMOR_HAS_GBM
+    if (strcmp(argv[i], "-vfbdevice") == 0) {
+        CHECK_FOR_REQUIRED_ARGUMENTS(1);
+        vfbDeviceNode = argv[++i];
+        return 2;
+    }
+#endif
+
     return 0;
 }
 
@@ -727,9 +754,108 @@ vfbCloseScreen(ScreenPtr pScreen)
         (*pScreen->DestroyPixmap) (pScreen->devPrivate);
     pScreen->devPrivate = NULL;
 
+#ifdef GLAMOR_HAS_GBM
+    if (pvfb->fd >= 0) {
+        if (pvfb->front_bo) {
+            gbm_bo_destroy(pvfb->front_bo);
+            pvfb->front_bo = NULL;
+        }
+        if (pvfb->gbm) {
+            gbm_device_destroy(pvfb->gbm);
+            pvfb->gbm = NULL;
+        }
+        close(pvfb->fd);
+        pvfb->fd = -1;
+    }
+#endif
+
     return pScreen->CloseScreen(pScreen);
 }
 
+#ifdef GLAMOR_HAS_GBM
+static Bool
+vfbCreateScreenResources(ScreenPtr pScreen)
+{
+    vfbScreenInfoPtr pvfb = &vfbScreens[pScreen->myNum];
+    PixmapPtr pixmap;
+    Bool ret;
+
+    pScreen->CreateScreenResources = pvfb->createScreenResources;
+    ret = pScreen->CreateScreenResources(pScreen);
+    pScreen->CreateScreenResources = vfbCreateScreenResources;
+
+    if (!ret)
+        return FALSE;
+
+    pixmap = pScreen->GetScreenPixmap(pScreen);
+
+    /* We don't support modifiers for the screen pixmap */
+    if (!glamor_egl_create_textured_pixmap_from_gbm_bo(pixmap, pvfb->front_bo,
+                                                       FALSE)) {
+        LogMessage(X_ERROR, "glamor_egl_create_textured_pixmap_from_gbm_bo() failed\n");
+        /* The screen is unusable, but there's no good way to unwind here */
+        return FALSE;
+    }
+
+    return TRUE;
+}
+
+
+static void
+vfbDRIInit(ScreenPtr pScreen)
+{
+    vfbScreenInfoPtr pvfb = &vfbScreens[pScreen->myNum];
+    const char *dri_node = vfbDeviceNode;
+    const char *error_msg = NULL;
+
+    pvfb->fd = open(dri_node, O_RDWR | O_CLOEXEC);
+    if (pvfb->fd < 0) {
+        error_msg = "Failed to open DRI render node";
+        goto fail;
+    }
+
+    if (!glamor_egl_init(pScreen, pvfb->fd)) {
+        error_msg = "Failed to initialize glamor EGL";
+        goto fail_fd;
+    }
+
+    pvfb->gbm = glamor_egl_get_gbm_device(pScreen);
+    if (!pvfb->gbm) {
+        error_msg = "Failed to get gbm device";
+        goto fail_fd;
+    }
+
+    pvfb->front_bo = gbm_bo_create(pvfb->gbm,
+                                   pScreen->width, pScreen->height,
+                                   GBM_FORMAT_ARGB8888,
+                                   GBM_BO_USE_RENDERING);
+    if (!pvfb->front_bo) {
+        error_msg = "Failed to create front buffer";
+        goto fail_fd;
+    }
+
+    if (!glamor_init(pScreen, GLAMOR_USE_EGL_SCREEN)) {
+        error_msg = "Failed to initialize glamor";
+        goto fail_bo;
+    }
+
+    pvfb->createScreenResources = pScreen->CreateScreenResources;
+    pScreen->CreateScreenResources = vfbCreateScreenResources;
+    LogMessage(X_INFO, "glamor/DRI3 initialized for VFB screen %d on %s\n", pScreen->myNum, dri_node);
+    return;
+
+ fail_bo:
+    gbm_bo_destroy(pvfb->front_bo);
+    pvfb->front_bo = NULL;
+ fail_fd:
+    close(pvfb->fd);
+    pvfb->fd = -1;
+ fail:
+    if (error_msg)
+        LogMessage(X_ERROR, "%s. Disabling GLAMOR/DRI3.\n", error_msg);
+}
+#endif
+
 static Bool
 vfbRROutputValidateMode(ScreenPtr           pScreen,
                         RROutputPtr         output,
@@ -922,12 +1044,15 @@ vfbScreenInit(ScreenPtr pScreen, int argc, char **argv)
 
     ret = fbScreenInit(pScreen, pbits, pvfb->width, pvfb->height,
                        dpix, dpiy, pvfb->paddedWidth, pvfb->bitsPerPixel);
-    if (ret && Render)
-        fbPictureInit(pScreen, 0, 0);
-
     if (!ret)
         return FALSE;
-
+    if (Render) {
+        fbPictureInit(pScreen, 0, 0);
+#ifdef GLAMOR_HAS_GBM
+        if (vfbDeviceNode)
+            vfbDRIInit(pScreen);
+#endif
+    }
     if (!vfbRandRInit(pScreen))
        return FALSE;
 
diff --git a/hw/vfb/Makefile.am b/hw/vfb/Makefile.am
index a4b4526..725ffca 100644
--- a/hw/vfb/Makefile.am
+++ b/hw/vfb/Makefile.am
@@ -3,7 +3,11 @@ SUBDIRS = man
 bin_PROGRAMS = Xvfb
 
 AM_CFLAGS = -DHAVE_DIX_CONFIG_H \
-            $(XVFBMODULES_CFLAGS) \
+            $(GBM_CFLAGS) \
+            $(DRM_CFLAGS) \
+	    -I$(top_srcdir)/glamor \
+	    -DGLAMOR_HAS_GBM=1 \
+	    $(XVFBMODULES_CFLAGS) \
 	    $(DIX_CFLAGS)
 
 SRCS =	InitInput.c \
@@ -14,13 +18,17 @@ SRCS =	InitInput.c \
 Xvfb_SOURCES = $(SRCS)
 
 XVFB_LIBS = \
-        @XVFB_LIBS@ \
+	$(top_builddir)/glamor/libglamor.la \
+	$(top_builddir)/glamor/libglamor_egl.la \
+	@XVFB_LIBS@ \
 	$(MAIN_LIB) \
 	$(XSERVER_LIBS) \
 	$(top_builddir)/Xi/libXistubs.la
 
-Xvfb_LDADD = $(XVFB_LIBS) $(XVFB_SYS_LIBS) $(XSERVER_SYS_LIBS)
-Xvfb_DEPENDENCIES = $(XVFB_LIBS)
+Xvfb_LDADD = $(XVFB_LIBS) $(XVFB_SYS_LIBS) $(XSERVER_SYS_LIBS) $(GBM_LIBS) $(DRM_LIBS) $(EPOXY_LIBS)
+Xvfb_DEPENDENCIES = \
+	$(top_builddir)/glamor/libglamor.la \
+	$(top_builddir)/glamor/libglamor_egl.la
 Xvfb_LDFLAGS = $(LD_EXPORT_SYMBOLS_FLAG)
 
 relink:
diff --git a/hw/vfb/meson.build b/hw/vfb/meson.build
index 7332866..de56e86 100644
--- a/hw/vfb/meson.build
+++ b/hw/vfb/meson.build
@@ -5,11 +5,18 @@ srcs = [
     '../../mi/miinitext.h',
 ]
 
+glamor_deps = []
+glamor_link = []
+if get_option('glamor') == 'true'
+    glamor_deps += dependency('gbm')
+    glamor_link += [glamor, glamor_egl]
+endif
+
 xvfb_server = executable(
     'Xvfb',
     srcs,
     include_directories: inc,
-    dependencies: common_dep,
+    dependencies: [common_dep] + glamor_deps,
     link_with: [
         libxserver_main,
         libxserver_fb,
@@ -18,7 +25,7 @@ xvfb_server = executable(
         libxserver_xi_stubs,
         libxserver_glx,
         libglxvnd,
-    ],
+    ] + glamor_link,
     install: true,
 )
 

